From eb0dc73b9853be7439a91b93c2b993fcfb6999e7 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 20 Dec 2024 11:28:44 +0100
Subject: [PATCH 798/841] fix(stm32mp2): support cortex M33 as TDCID in
 STM32MP25 low power

When Cortex-M33 is TDCID, Cortex-A35 and TF-A are not supposed to
initialize or modify CPU2 related power section.

Fix BL31 PM support for cortex A35 when cortex M33 is TDCID (boot device)
- only support Stop2 mode: LP or LPLV is selected by M33
  for PSCI system suspend
- remove access to PWR_CPU2CR register
- for power OFF, just request Standby, remove access to TDCID access
  register (RCC_CPUBOOTCR, RCC_C2RSTCSETR, PWR_CR10, ...)

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 17d5fcdfa223d5dc5e4a6cd41378827b8dd35fe7 fix(stm32mp2): support cortex M33 as TDCID in STM32MP25 low power]
Fixes: 0fca5ccfdc61 ("feat(stm32mp2): support cortex M33 as TDCID in STM32MP25 low power")
Change-Id: I75c1ebd7a35a19648a0049a90e2c6135107f1fb7
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/426157
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_def.h |  2 +
 plat/st/stm32mp2/stm32mp2_pm.c  | 70 +++++++++++++++++++++++++++------
 2 files changed, 59 insertions(+), 13 deletions(-)

diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 1f02a799a..3c9544c7e 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -352,6 +352,8 @@ enum ddr_type {
 
 #define EXTI1_C1IMR2_WKUP_MASK		GENMASK_32(57 - 32, 52 - 32)
 
+#define EXTI1_C1IMR3_CPU2_SEV		BIT_32(64 - 64)
+
 /* EXTI1 software interrupt event register  */
 #define EXTI1_SWIER3			U(0x48)
 #define EXTI1_RPR3			U(0x4C)
diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 5d2b64bbd..66ac31d3c 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -265,7 +265,7 @@ static void stm32mp_ca35_lpi_restore(void)
 /* LPSRAM1/2/3 autonomous support (AMEN) when D3 is used in low power */
 static void lpsram_autonomous_mode_set(uintptr_t rcc_base)
 {
-        /* STM32MP21 Series don't have D3 domain */
+	/* STM32MP21 Series don't have D3 domain */
 #if !STM32MP21 && !STM32MP_M33_TDCID
 	const uint32_t mask = RCC_C3CFGR_C3EN | RCC_C3CFGR_C3AMEN;
 
@@ -466,14 +466,12 @@ static bool stm32_freeze_other_core(unsigned int core_id)
 	return result;
 }
 
-static bool stm32_pwr_cpu2_state_is_running(__maybe_unused uintptr_t pwr_base)
+#if !STM32MP_M33_TDCID
+static bool stm32_pwr_cpu2_state_is_running(uintptr_t pwr_base)
 {
-#if STM32MP_M33_TDCID
-	return true;
-#else
 	return (mmio_read_32(pwr_base + PWR_CPU2D2SR) & PWR_CPU2D2SR_CSTATE_MASK) != 0U;
-#endif
 }
+#endif
 
 #if !STM32MP_M33_TDCID
 static bool stm32_pwr_cpu3_state_is_running(__maybe_unused uintptr_t pwr_base)
@@ -488,7 +486,6 @@ static bool stm32_pwr_cpu3_state_is_running(__maybe_unused uintptr_t pwr_base)
 
 static int stm32_pwr_domain_validate_suspend(const psci_power_state_t *target_state)
 {
-	uintptr_t pwr_base = stm32mp_pwr_base();
 	uint32_t stateid = stm32_get_stateid(target_state->pwr_domain_state);
 	u_register_t mpidr = read_mpidr();
 	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
@@ -498,8 +495,15 @@ static int stm32_pwr_domain_validate_suspend(const psci_power_state_t *target_st
 		return PSCI_E_SUCCESS;
 	}
 
+#if STM32MP_M33_TDCID
+	/* CA35 can't choose LP-Stop2 or LPLV-Stop2 modes when M33 TDCID */
+	if (stateid == PWRSTATE_LP_STOP2 || stateid == PWRSTATE_LPLV_STOP2) {
+		WARN("Invalid PSCI power state %x with Cortex M33 TDCID.\n", stateid);
+		return PSCI_E_INVALID_PARAMS;
+	}
+#else
 	/* If CPU2 is not in reset: limit supported low power modes */
-	if (stm32_pwr_cpu2_state_is_running(pwr_base)) {
+	if (stm32_pwr_cpu2_state_is_running(stm32mp_pwr_base())) {
 		/* PMIC update in OP-TEE is not allowed with M33 running */
 		if (stateid == PWRSTATE_STANDBY ||
 		    stateid == PWRSTATE_LPLV_STOP2 ||
@@ -508,6 +512,7 @@ static int stm32_pwr_domain_validate_suspend(const psci_power_state_t *target_st
 			return PSCI_E_INVALID_PARAMS;
 		}
 	}
+#endif
 
 	if (!stm32_freeze_other_core(core_id)) {
 		return PSCI_E_DENIED;
@@ -534,8 +539,10 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 	uintptr_t rcc_base = stm32mp_rcc_base();
 	bool standby __maybe_unused = false;
 	uint32_t stateid = stm32_get_stateid(target_state->pwr_domain_state);
+#if !STM32MP_M33_TDCID
 	bool cpu2_running = stm32_pwr_cpu2_state_is_running(pwr_base);
 	uint32_t pwr_r3cidcfgr = 0U;
+#endif
 
 	/* If retention only at D1 level return as nothing is to be done */
 	if (stateid == PWRSTATE_RUN) {
@@ -581,7 +588,6 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 
 	/* Disable DDRSHR to avoid STANDBY/STOP exit issue */
 	mmio_clrbits_32(rcc_base + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRSHR);
-#endif
 
 	if (!cpu2_running) {
 		/* Filtering for CID1 = CA35 on PWR resource 3 = PWR_CPU2CR */
@@ -589,25 +595,28 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 		mmio_write_32(pwr_base + PWR_R3CIDCFGR,
 			      (RIF_CID1 << PWR_R3CIDCFGR_SCID_SHIFT) | PWR_R3CIDCFGR_CFEN);
 	}
+#endif
 
 	/* Perform the PWR configuration for the requested mode */
 	switch (stateid) {
 	case PWRSTATE_STOP1:
 		print_mode_verbose("Stop1");
 		mmio_write_32(pwr_base + PWR_CPU1CR, 0U);
+#if !STM32MP_M33_TDCID
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR, 0U);
 		}
+#endif
 		stm32mp2_enable_rcc_wakeup_irq(rcc_base);
 		break;
 
 	case PWRSTATE_LP_STOP1:
 		print_mode_verbose("LP_Stop1");
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_LPDS_D1);
+#if !STM32MP_M33_TDCID
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR, PWR_CPU2CR_LPDS_D2);
 		}
-#if !STM32MP_M33_TDCID
 		/* Wait VTT ramp-up delay for LP-Stop1 */
 		mmio_write_32(rcc_base + RCC_PWRLPDLYCR, lpstop1_pwrlpdly);
 #endif
@@ -618,24 +627,29 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 		print_mode_verbose("LPLV_Stop1");
 		lpsram_autonomous_mode_set(rcc_base);
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_LPDS_D1 | PWR_CPU1CR_LVDS_D1);
+#if !STM32MP_M33_TDCID
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR,
 				      PWR_CPU2CR_LPDS_D2 | PWR_CPU2CR_LVDS_D2);
 		}
+#endif
 		stm32mp2_enable_rcc_wakeup_irq(rcc_base);
 		break;
 
 	case PWRSTATE_STOP2:
 		print_mode_info("Stop2");
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1);
+#if !STM32MP_M33_TDCID
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR, 0U);
 		}
+#endif
 		stm32mp_gic_cpuif_disable();
 		stm32mp_gic_save();
 		stm32mp2_pll1_disable();
 		break;
 
+#if !STM32MP_M33_TDCID
 	case PWRSTATE_LP_STOP2:
 		print_mode_info("LP_Stop2");
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1);
@@ -659,14 +673,17 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 		stm32mp_gic_save();
 		stm32mp2_pll1_disable();
 		break;
+#endif
 
 	case PWRSTATE_STANDBY:
 		print_mode_info("Standby1");
 		lpsram_autonomous_mode_set(rcc_base);
 		mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1 | PWR_CPU1CR_PDDS_D2);
+#if !STM32MP_M33_TDCID
 		if (!cpu2_running) {
 			mmio_write_32(pwr_base + PWR_CPU2CR, PWR_CPU2CR_PDDS_D2);
 		}
+#endif
 		stm32mp_gic_cpuif_disable();
 		stm32mp2_pll1_disable();
 		break;
@@ -680,11 +697,13 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 
 	/* Clear previous status */
 	mmio_setbits_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_CSSF);
+#if !STM32MP_M33_TDCID
 	if (!cpu2_running) {
 		mmio_setbits_32(pwr_base + PWR_CPU2CR, PWR_CPU2CR_CSSF);
 		/* Restore RIF config for resource 3 = PWR_CPU2CR */
 		mmio_write_32(pwr_base + PWR_R3CIDCFGR, pwr_r3cidcfgr);
 	}
+#endif
 
 #if !STM32MP21 && !STM32MP_M33_TDCID
 	mmio_setbits_32(pwr_base + PWR_CPU3CR, PWR_CPU3CR_CSSF);
@@ -901,8 +920,10 @@ static void __dead2 stm32_system_off(void)
 	u_register_t mpidr = read_mpidr();
 	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
 	uintptr_t exti2_base = STM32MP_EXTI2_BASE;
+#if !STM32MP_M33_TDCID
 	uint32_t otp_idx = 0U;
 	uint32_t otp_value = 0U;
+#endif
 
 	if (core_id != STM32MP_PRIMARY_CPU) {
 		ERROR("PSCI system off request on core %u\n", core_id);
@@ -931,6 +952,7 @@ static void __dead2 stm32_system_off(void)
 	}
 #endif /* !STM32MP21 */
 
+#if !STM32MP_M33_TDCID
 	/* If CPU2 is not in reset */
 	if (stm32_pwr_cpu2_state_is_running(pwr_base)) {
 		WARN("PSCI system off with Cortex M33 running.\n");
@@ -939,6 +961,7 @@ static void __dead2 stm32_system_off(void)
 		mmio_setbits_32(rcc_base + RCC_C2RSTCSETR, RCC_C2RSTCSETR_C2RST);
 		dsb();
 	}
+#endif
 
 #if !STM32MP21
 	/* If CPU3 is not in reset */
@@ -949,6 +972,7 @@ static void __dead2 stm32_system_off(void)
 	}
 #endif /* !STM32MP21 */
 
+#if !STM32MP_M33_TDCID
 	/* Freeze all watchdog with shadow value of HCONF1 */
 	if (stm32_get_otp_index(HCONF1_OTP, &otp_idx, NULL) == 0U) {
 		if (stm32_get_otp_value_from_idx(otp_idx, &otp_value) == 0U) {
@@ -959,6 +983,7 @@ static void __dead2 stm32_system_off(void)
 			stm32_otp_write(otp_value, otp_idx);
 		}
 	}
+#endif
 
 	dcsw_op_all(DCCISW);
 
@@ -973,19 +998,22 @@ static void __dead2 stm32_system_off(void)
 	/* Request STOP for both cores */
 	mmio_write_32(rcc_base + RCC_C1SREQSETR, RCC_C1SREQSETR_STPREQ_MASK);
 
-	/* Deactivate CID filtering on region 3 for PWR_CPU2CR */
-	mmio_write_32(pwr_base + PWR_R3CIDCFGR, 0U);
-
 	/* Request standby2 */
 	mmio_write_32(pwr_base + PWR_CPU1CR, PWR_CPU1CR_PDDS_D1 | PWR_CPU1CR_PDDS_D2);
+#if !STM32MP_M33_TDCID
+	/* Deactivate CID filtering on region 3 for PWR_CPU2CR */
+	mmio_write_32(pwr_base + PWR_R3CIDCFGR, 0U);
 	mmio_write_32(pwr_base + PWR_CPU2CR, PWR_CPU2CR_PDDS_D2);
+#endif
 #if !STM32MP21
 	mmio_write_32(pwr_base + PWR_D3CR, PWR_D3CR_PDDS_D3);
 #endif /* !STM32MP21 */
 	stm32mp2_pll1_disable();
 
+#if !STM32MP_M33_TDCID
 	/* Do not maintain RETRAM memory content in Standby or Vbat */
 	mmio_write_32(pwr_base + PWR_CR10, PWR_CR10_RETRBSEN_DISABLE);
+#endif
 
 	/* Clear PM context in BKPSRAM: cold boot at next wake-up */
 	stm32_pm_context_clear();
@@ -1168,6 +1196,15 @@ static void stm32_get_sys_suspend_power_state(psci_power_state_t *req_state)
 	uint32_t c1imr3 = mmio_read_32(exti_base + EXTI1_C1IMR3);
 
 	/* Verify the max level supported according to the activated EXTI1 */
+#if STM32MP_M33_TDCID
+	/* CM33 selects the Stop2 mode (LP or LPLV), only check Standby1 restrictions */
+	if ((c1imr1 != 0U) || ((c1imr2 & ~EXTI1_C1IMR2_WKUP_MASK) != 0U) ||
+	    ((c1imr3 & ~EXTI1_C1IMR3_CPU2_SEV) != 0U)) {
+		max_pwr_state = PWRSTATE_STOP2;
+		VERBOSE("%s: max_pwr_state = PWRSTATE_LP_STOP2, C1IMR1=%x, C1IMR2=%x, C1IMR3=%x\n",
+			__func__, c1imr1, c1imr2, c1imr3);
+	}
+#else
 	if ((c1imr1 & (EXTI1_C1IMR1_GPIO | EXTI1_C1IMR1_PVD | EXTI1_C1IMR1_PVM)) != 0U) {
 		max_pwr_state = PWRSTATE_LPLV_STOP2;
 		VERBOSE("%s: max_pwr_state = PWRSTATE_LPLV_STOP2 C1IMR1=%x\n", __func__, c1imr1);
@@ -1187,6 +1224,7 @@ static void stm32_get_sys_suspend_power_state(psci_power_state_t *req_state)
 		VERBOSE("%s: max_pwr_state = PWRSTATE_LP_STOP2, M33 is running\n", __func__);
 	}
 
+#endif
 	/* Trace to debug low power mode restriction */
 	if (max_pwr_state != PWRSTATE_STANDBY) {
 		INFO("max_pwr_state=%x C1IMR1=%x C1IMR2=%x C1IMR3=%x CPU2D2SR=%x\n",
@@ -1276,6 +1314,12 @@ static int stm32_parse_domain_idle_state(void *fdt)
 			return -EINVAL;
 		}
 
+#if STM32MP_M33_TDCID
+		/* LP-Stop2 or LPLV-Stop2 not managed by CA35 for CM33 TDCID */
+		if (power_state == PWRSTATE_LP_STOP2 || power_state == PWRSTATE_LPLV_STOP2)
+			continue;
+#endif
+
 		domain_idle_states[i++] = power_state;
 
 		/* Check array size */
-- 
2.47.2

